### ISSUE TITLE
Drop the use of `-coreid` in RISC-V targets

### ISSUE THREAD

**Issue by en-sc** [riscv/riscv-openocd#1147]
Currently `-coreid` is used in RISC-V targets as `hartid`.
https://github.com/riscv-collab/riscv-openocd/blob/a4020f1a0207dc218812ee262ce861d81aa7962f/src/target/riscv/riscv-013.c#L1909
This is misleading.

(`hartid` name is probably not the best. Here it's used just as an example.)

1. ~By default `hartid` can be assigned based on target index~ See https://github.com/riscv-collab/riscv-openocd/issues/1147#issuecomment-2416145221
2. RISC-V-specific `-hartid` configure option can be introduced.
3. Configuring `-coreid` on a RISC-V target should have the same effect as `-hartid` and a deprecation message should be shown.

**Comment by JanMatCodasip** [comment]
Hi @en-sc 

I agree that the name `-coreid` is misleading in case of RISC-V.

> By default hartid can be assigned based on target index.

I am not sure I understand this fully. Where would the hard index be taken from if `-coreid` is removed?

> RISC-V-specific -hartid configure option can be introduced.

That'd be my preference. Furthermore:

- I'd pick a different name for the parameter to not confuse the user that it has something to do with the value of `mhartid` CSR register. It could be named for example `-hart_index <num>` or `-hart <num>` or similar.
- It'd be good to still support `-coreid <num>` but show a deprecation warning, allowing the users to adapt their TCL scripts.

**Comment by en-sc** [comment]
> It'd be good to still support -coreid <num> but show a deprecation warning

You are right. Updated the description.

> I am not sure I understand this fully. Where would the hard index be taken from if -coreid is removed?

From the order the targets are created in:
```
target create ... # hartid implicitly 0
target create ... # hartid implicitly 1
```

IMHO it's a nice default, considering that the targets are usually the same anyway (SMP) (especially when they are connected to the same DM).

Moreover, this will make validation of `hartid` unnecessary (it should be contiguous and start from zero).

And there is also the case of a platform with multiple DMs (here `hartid` is not unique).

**Comment by JanMatCodasip** [comment]
> Where would the hard index be taken from if -coreid is removed?
> > From the order the targets are created in

I'd strongly prefer to not have implicitly incrementing hart numbers for these reasons:
- For clarity ("explicit is better than implicit" - PEP 20)
- To not change the behavior: now omitted hart ID means fixed `0`, after the change the omitted hart ID would behave very differently
- For various (less or more obscure) reasons the user may not want to debug all harts belonging to a given debug module. For example the user may only be interested in debugging harts 0 and 3.

Furthermore, it'd be good to check the uniqueness of hart number for a given debug module (that is, for a given pair of JTAG TAP and `-dbgbase`).

**Comment by en-sc** [comment]
I see your point, however I'd like to disagree:

> To not change the behavior: now omitted hart ID means fixed 0, after the change the omitted hart ID would behave very differently

This is not true. AFAIU, if harts with omitted hart ID are on the same DM -- it's an error that is not currently diagnosed.
If hart ID is specified with a value greater then the detected number of harts on the DM --it's also an error that is not diagnosed.
(it sometimes "works" since `hartsel` is WARL). What I'm trying to say is, IMHO there is not much "behavior" to preserve here.

> For various (less or more obscure) reasons the user may not want to debug all harts belonging to a given debug module. For example the user may only be interested in debugging harts 0 and 3.

AFAIU, this is a non-default scenario which can be covered by non-default config.

> For clarity ("explicit is better than implicit" - PEP 20)

Oh why did you have to quote PEP here.. I'd agree with the argument but I hate Python with all my heart :)

**Comment by en-sc** [comment]
In all seriousness though, my main argument is: assigning `hartid` based on target index makes it correct by default most of the time. Otherwise extensive verification is required.

E.g.a system with two DMs on one TAP:
* Explicit hart ID:|
```
target create ... -dbgbase 0 -hartid 0
target create ... -dbgbase 0 -hartid 1
target create ... -dbgbase yyy -hartid 0
target create ... -dbgbase yyy -hartid 1
```
* Implicit hart ID:
```
target create ...  # -dbgbase 0, hartid 0
target create ...
target create ... # -dbgbase based on the value of `nextdm`, hartid 0
target create ...
```

You can see that requiring the user to specify `hartid` explicitly makes it so the user needs to specify information that can be easily examined (and should be examined to verify that user configuration is correct).

**Comment by JanMatCodasip** [comment]
> me: To not change the behavior: now omitted hart ID means fixed 0, after the change the omitted hart ID would behave very differently

> > en-sc: This is not true. AFAIU, if harts with omitted hart ID are on the same DM -- it's an error that is not currently diagnosed. (...)  IMHO there is not much "behavior" to preserve here.

Fair point, I agree! 

I also am in agreement that OpenOCD should auto-detect the base addresses of all DMs and the number of harts on each DM.

> me: For various (less or more obscure) reasons the user may not want to debug all harts (...).

> > en-sc: AFAIU, this is a non-default scenario which can be covered by non-default config.

I would like there to not be two ways of writing the configuration (default vs. non-default, or we could call them automatic vs. explicit). 

In my opinion, the only advantage of the "automatic" configuration is shorter configuration - less characters to type.

Other than that, I am afraid that I can only see disadvantages:

- Too much automation happening behind the scenes (dbgbase and coreid automatically assigned).
- Two ways of writing the same thing: automatic vs. explicit.
- How would automatic and explicit configuration work together? What if just some -hartid and -dbgbase items are specified and others left blank. How would they interact? Would it behave like implicitly vs. explicitly defined values in C enums? Or differently? The chosen behavior would also all need to be described clearly in the documentation. The users could have problems understanding the meaining of the configs without reading that documentation.
- Structure of the system will not be clear after reading the configuration file.
- It won't be immediately apparent to the user that -dbgbase and -coreid options actually exist just by looking at the configuration file.
- Increased likelihood of mistakes: Configuration file for one SoC could be by mistake applied onto a different SoC with different structure but same number of harts - the user may not notice.
- Automatic configuration would not follow the principles that OpenOCD already uses in similar places: Take the JTAG chain configuration as an example. OpenOCD is able to auto-probe the JTAG chain but still requires the user to specify the expected chain structure.

For those reasons, I'd prefer to not have the automatic configuration. 

To summarize, my preference would be:

- Keep the current defaults only: `-dbgbase 0x0`, `-hartid 0`. Do not try to make OpenOCD more clever. The current defaults nicely cover the most basic case of single DM and single hart.
- Ask the user to explicitly specify dbgbase and hartid for any case that is more complex than the above.
- Display error messages on obvious mistakes in the configuration prior to examine() (e.g. duplicate pairs of dbgbase and hartid).
- Detect the number of DMs and harts in the hardware during examine().
- Validate the user configuration against what has been detected.
- If there is a mismatch, provide a hint to the user what configuration commands they should use - in the same manner as JTAG chain auto-probing works.

**Comment by lz-bro** [comment]
Hi, @en-sc @JanMatCodasip 
Currently, `-dbgbase` is used to configure the base address of DM. I would like to introduce RISC-V specific configuration like `-dm-num` and determine the base address of DM through `nextdm`, so that OpenOCD configuration can be more flexible. In addition, due to some special scenarios, such as when a DM power down, we cannot obtain the `nextdm` through the power-off DM, but we still need to debug the harts on other power-on DMs. So I tend to keep `-dbgbase`, when `target->dbgbase_set` is true, we don't need to check whether this dbgbase exists through nextdm.

**Comment by JanMatCodasip** [comment]
Hi @lz-bro,

just double-checking if I understood you right. Did you mean this?

```
-dm-num <N>  ---> Find the N-th debug module by inspecting the chain of addresses in the nexdm
                  debug module registers.

-dbgbase <ADDR>  ---> OpenOCD would just use the debug module address as given by the user
                      and not check whether that DM really exists.
```

If yes, that sounds reasonable. 

However, I don't like that with `-dm-num`, the debug module address is gone from the config file, and the configuration that way becomes less explicit for the readers.

As an alternative, you can consider something like this `-dbgbase <ADDR> -skip-dm-discovery`.

In any case, if you are preparing a patch, please post it to review.openocd.org, not here.

**Comment by lz-bro** [comment]
Hi @JanMatCodasip 

> just double-checking if I understood you right. Did you mean this?

yes.

> However, I don't like that with -dm-num, the debug module address is gone from the config file, and the configuration that way becomes less explicit for the readers.

I think that users might not be able to immediately get the base address of the target DM. Using the -dm-num option, it would be easier for them to write the configuration scripts they want.

> As an alternative, you can consider something like this -dbgbase <ADDR> -skip-dm-discovery.

I already have related alternatives, like `-check-dbgbase <on|off>`, but I still want to confirm whether I should include the `-dm-num` option

**Comment by JanMatCodasip** [comment]
> I think that users might not be able to immediately get the base address of the target DM.

What use cases do you have in mind where the users would not know the base address of the DM in the platform at the time they write their OpenOCD configuration file? Are the debug module base addresses not fixed for a given piece of hardware?

> I still want to confirm whether I should include the -dm-num option

I am still little uneasy about this proposed option. I'd prefer the users to be explicit about the configuration of their target - similarly as they specify the number and names of targets, SMP groups etc.

**Comment by en-sc** [comment]
I'd like to share my perspective on this. In an ideal world, an OpenOCD user (the person writing the config) should be able to achieve their goals without needing any knowledge of the RISC-V Debug Specification. To me, the ideal configuration would look like this:

```
jtag newtap rv_dtm0 tap -irlen 5
jtag newtap rv_dtm1 tap -irlen 5
...
riscv_dtm create rv_dtm0 -chain-position rv_dtm0
...
target create c0 riscv -dtm rv_dtm0 -mhartid 0
target create c1 riscv -dtm rv_dtm0 -mhartid 1
```

This approach provides two main benefits:

   1. Treating the DTM as a separate entity simplifies both the examination process and the target hierarchy.
   1. Identifying the hart by its `mhartid` leverages a concept users are already familiar with.

> What use cases do you have in mind where the users would not know the base address of the DM in the platform at the time they write their OpenOCD configuration file? Are the debug module base addresses not fixed for a given piece of hardware?

Requiring the base address is a poor user experience. When writing a configuration for a poorly documented chip, finding the DM address requires a lot of guesswork since it is not something a user can easily observe.

That said, we should also provide options to minimize the invasiveness and execution time of the examination procedure. This is especially helpful in simulation environments or when hot-plugging to a target after an issue has already occurred.
To solve this, we could offer three configuration paths for any given parameter:

   1. Auto-discovery: The parameter (e.g., DM address or supported extensions) is automatically discovered during examination.
   1. User-override: The parameter is explicitly provided by the user in the configuration, bypassing discovery.
   1. Validation: The user provides the parameter, and OpenOCD validates it during examination.

To keep the interface consistent, we could pass these user-provided overrides as a dictionary and use a separate array to specify which parameters to validate during examination. By default all the parameters are discovered. For example:
```
# By default everything is discovered
target create c0 ...
# DM address is explicitly provided by the user, not validated.
target create c0 ... -riscv-config dm-addr 0x1234
# The user-provided DM address is validated against the discovered value.
target create c0 ... -riscv-config dm-addr 0x1234 -riscv-validate dm-addr
```

**Comment by lz-bro** [comment]
> 1. Treating the DTM as a separate entity simplifies both the examination process and the target hierarchy.

I agree with treating the DTM as a separate entity, and the DTM should be able to detect the number and base addresses of DMs, as well as the number of harts under each DM.  Therefore, we should need to configure the index representing DM, such as `-dm-num`, the base address of DM, such as `-dbgbase`, or `-dm-base`, the index of hart, such as `-coreid`, or `-hart-num,` etc. Additionally, I do not have a strong intention to change the generic `-dbgbase` and `-coreid` to RISC-V spec related names.

> 2. Identifying the hart by its `mhartid` leverages a concept users are already familiar with.

I don't agree with this point. Using `mhartid`, I doubt whether we can accurately confirm the DM and hart index to which the target belongs. Also, if we do confirm, do we still need to verify mhartid CSR? I think this makes the situation more complicated.

Based on the above, I think that the configuration would look like this:

E.g.a system with two DMs on one DTM:

```
jtag newtap dtm0 tap -irlen 5

dtm create dtm0 -chain-position dtm0

target create dm0.c0 riscv -dtm dtm0 -dm-num 0 -coreid 0  # or -dbgbase 0
target create dm0.c1 riscv -dtm dtm0 -dm-num 0 -coreid 1  # or -dbgbase 0
target create dm1.c0 riscv -dtm dtm0 -dm-num 1 -coreid 0  # or -dbgbase xxx
target create dm1.c1 riscv -dtm dtm0 -dm-num 1 -coreid 1  # or -dbgbase xxx

```

>To keep the interface consistent, we could pass these user-provided overrides as a dictionary and use a separate array to specify >which parameters to validate during examination. By default all the parameters are discovered.

In my opinion, there is no need to introduce two parameters to determine whether dbgbase and coreid need to be checked, because when dbgbase cannot be obtained through nextdm, the check of coreid has actually expired, so we can only use one parameter

**Comment by JanMatCodasip** [comment]
Hi @en-sc and @lz-bro,

thank you for your replies. I can see that this will be a controversial topic as I also have somewhat different view on it. :-)

> en-sc: In an ideal world, an OpenOCD user (the person writing the config) should be able to achieve their goals without needing any knowledge of the RISC-V Debug Specification.

I am sorry to disagree. I believe that the configuration file should use the identifiers designated by the RISC-V debug specification for that purpose - that is, debug module addresses and hart numbers (not mhartid, not ordinal number of the debug module ...).

Furthermore, I believe the configuration files should be explicit - they should contain the identifiers of the debug components explicitly written down, so that:

- the intent of the author is clearly captured there,
- the author's intent can later be reviewed by someone else (e.g. when troubleshooting),
- future users can learn about the hardware structure by reading the configuration file.

> en-sc: Requiring the base address is a poor user experience. When writing a configuration for a poorly documented chip, finding the DM address requires a lot of guesswork since it is not something a user can easily observe.

I would prefer to address use cases like this by a form of "autoprobing": In this special mode, OpenOCD would detect all debug modules, harts connected to them as well as their mhartid, in a best-effort way, and print all this information to the user in a concise form.

The users can then leverage this information (along with other knowledge or documentation of the target hardware) to write a concrete, explicit configuration file.

The JTAG chain auto-probing - a mechanism which is already part of OpenOCD - works this way.  OpenOCD can probe the JTAG chain for the user, but then the user still needs to be concrete about what the chain looks like in the configuration file.

>  en-sc: To solve this, we could offer three configuration paths for any given parameter: (...)

I am fine with a way how to skip the validation of any parameter as long as the validation remains enabled by default (and the opt-out is explicit user action) - so that mistakes are captured early and users' time is not wasted.

> en-sc: Treating the DTM as a separate entity simplifies both the examination process and the target hierarchy (...)

I am not against having RISC-V DTM as a separate entity in the config.

You could even have a RISC-V DM as a separate entity. But I have not thought too deeply about what it would mean and what the benefits would be.

**Comment by borneoa** [comment]
Please also consider the SoC where the RISC-V cores are behind a ARM DAP+AP, not behind DTM.
I know there is some activity to add to OpenOCD RISC-V behind DAP, I have not checked it yet.
And I should read again the debug spec of RISC-V. Sorry if there is something not fully correct below.

In OpenOCD, on ARM CPUs behind DAP, we have a similar problem to parse some table.
There we have:
- one or more DAP (equivalent to the DTM here ?)
- behind the DAP there are the APs (not present here, even if we should have it on RISC-V in DAP+AP based SoC)
- behind the AP there is one or more CPUs plus one or more ROM tables (similar to the nextdm list)

OpenOCD has dedicated commands to dump the ROM tables, so on new SoC we can initially create a simple configuration file that only gives access to the DAP and use it to dump the ROM table to find the CPUs. Once found the base addresses of the CPUs, we can write the full config script for that SoC.

OpenOCD is also able to parse the ROM table at examine and select the N'th CPU, where N is specified by `-coreid`. But accessing and parsing is slow, plus the ROM could be bugged or partially not-accessible due to low-power. The `-dbgbase` is used to hardcode the address, bypassing the parsing; no checking is done in this case.

It would be nice to have something similar for RISC-V too.
- a new command `drm info <dm_num>` or similar that parses one DM and dumps what it can find inside
- at target create either hardcoding `-dbgbase` or using some index to auto detect the nth CPU.

**Comment by lz-bro** [comment]
Hi all, 
I referred to the practices of ARM DAP to try to solve these problems, I've posted my patch on review.openocd.org, please see [target/riscv: add dtm support](https://review.openocd.org/c/openocd/+/9695).
